flink-table/flink-table-api-java/src/test/java/org/apache/flink/table/module/ModuleManagerTest.java

```diff
@@ -0,0 +1,239 @@
+    }
+
+    @Test
+    public void testUnloadModuleTwice() {
+        assertEquals(Collections.singletonList(MODULE_TYPE_CORE), manager.getUsedModules());
+
+        manager.unloadModule(MODULE_TYPE_CORE);
+        assertEquals(Collections.emptyList(), manager.getUsedModules());
+        assertEquals(Collections.emptyMap(), manager.getLoadedModules());
+
+        thrown.expect(ValidationException.class);
+        thrown.expectMessage("No module with name core exists");
+        manager.unloadModule(MODULE_TYPE_CORE);
+    }
+
+    @Test
+    public void testUseUnloadedModules() {
+        thrown.expect(ValidationException.class);
+        thrown.expectMessage("No module with name x exists");
+        manager.useModules(MODULE_TYPE_CORE, "x");
+    }
+
+    @Test
+    public void testUseModulesWithDuplicateModuleName() {
+        thrown.expect(ValidationException.class);
+        thrown.expectMessage("Module core appears more than once");
+        manager.useModules(MODULE_TYPE_CORE, MODULE_TYPE_CORE);
+    }
+
+    @Test
+    public void testUseModules() {
+        ModuleMock.ModuleX x = new ModuleMock.ModuleX("x");
+        ModuleMock.ModuleY y = new ModuleMock.ModuleY("y");
+        ModuleMock.ModuleZ z = new ModuleMock.ModuleZ("z");
+        manager.loadModule(x.getType(), x);
+        manager.loadModule(y.getType(), y);
+        manager.loadModule(z.getType(), z);
+
+        assertEquals(Arrays.asList(MODULE_TYPE_CORE, "x", "y", "z"), manager.getUsedModules());
+
+        // test order for used modules
+        manager.useModules("z", MODULE_TYPE_CORE);
+        assertEquals(Arrays.asList("z", MODULE_TYPE_CORE), manager.getUsedModules());
+
+        // test unmentioned modules are still loaded
+        Map<String, Module> expectedLoadedModules = new HashMap<>();
+        expectedLoadedModules.put(MODULE_TYPE_CORE, CoreModule.INSTANCE);
+        expectedLoadedModules.put("x", x);
+        expectedLoadedModules.put("y", y);
+        expectedLoadedModules.put("z", z);
+        assertEquals(expectedLoadedModules, manager.getLoadedModules());
+    }
+
+    @Test
+    public void testListModules() {
+        ModuleMock.ModuleY y = new ModuleMock.ModuleY("y");
+        ModuleMock.ModuleZ z = new ModuleMock.ModuleZ("z");
+        manager.loadModule("y", y);
+        manager.loadModule("z", z);
+        manager.useModules("z", "y");
+
+        assertEquals(Arrays.asList("z", "y"), manager.listModules());
+    }
+
+    @Test
+    public void testListFullModules() {
+        ModuleMock.ModuleZ x = new ModuleMock.ModuleZ("x");
+        ModuleMock.ModuleY y = new ModuleMock.ModuleY("y");
+        ModuleMock.ModuleZ z = new ModuleMock.ModuleZ("z");
+
+        manager.loadModule("y", y);
+        manager.loadModule("x", x);
+        manager.loadModule("z", z);
+        manager.useModules("z", "y");
+
+        assertEquals(
+                getExpectedModuleEntries(2, "z", "y", MODULE_TYPE_CORE, "x"),
+                getActualModuleEntries());
```

Comment:
What do you think about making `ModuleManager.loadedModules` to be `LinkedHashMap`? This can provide deterministic result for `listFullModules` and make the testing easier. 

If we do this, we should add a comment on the `loadedModules` to explain why we use `LinkedHashMap`.

Reply:
+1 for generating the deterministic result.